File: backend/internal/shared/jobs/jobs_test.go
package jobs

import "testing"

func TestBuildEmailDispatchPayload(t *testing.T) {
	t.Run("rejects redacted fields", func(t *testing.T) {
		for _, field := range []string{"rawEmailCode", "emailVerificationUrl", "recipientEmail", "emailBody"} {
			_, err := BuildEmailDispatchPayload(map[string]string{
				"authChallengeId": "challenge-1",
				field:             "secret",
			})
			if err == nil {
				t.Fatalf("BuildEmailDispatchPayload allowed redacted field %s", field)
			}
		}
	})

	t.Run("accepts allowed fields", func(t *testing.T) {
		payload, err := BuildEmailDispatchPayload(map[string]string{
			"authChallengeId":   "challenge-1",
			"userId":            "user-1",
			"templateKey":       "login",
			"locale":            "en-US",
			"deliverySecretRef": "secret-ref-1",
			"dedupeKey":         "dedupe-1",
		})
		if err != nil {
			t.Fatalf("BuildEmailDispatchPayload: %v", err)
		}
		if payload["deliverySecretRef"] != "secret-ref-1" {
			t.Fatalf("payload did not preserve allowed field: %#v", payload)
		}
	})
}

func TestGeneratedJobMappings(t *testing.T) {
	if AsynqTaskTargetImport != "target.import" {
		t.Fatalf("AsynqTaskTargetImport = %q", AsynqTaskTargetImport)
	}
	if AsynqTaskPrivacyDelete != "privacy.delete" {
		t.Fatalf("AsynqTaskPrivacyDelete = %q", AsynqTaskPrivacyDelete)
	}
	if AsynqTaskEmailDispatch != "email.dispatch" {
		t.Fatalf("AsynqTaskEmailDispatch = %q", AsynqTaskEmailDispatch)
	}
	if JobTriggerEventSemanticSourceEventOnly != "source_event_only" {
		t.Fatalf("JobTriggerEventSemanticSourceEventOnly = %q", JobTriggerEventSemanticSourceEventOnly)
	}
	if JobTriggerEventSemanticTriggerCreatesJob != "trigger_creates_job" {
		t.Fatalf("JobTriggerEventSemanticTriggerCreatesJob = %q", JobTriggerEventSemanticTriggerCreatesJob)
	}
	if JobTriggerEventSemantics[JobTypeReportGenerate] != JobTriggerEventSemanticSourceEventOnly {
		t.Fatalf("report_generate semantic = %q", JobTriggerEventSemantics[JobTypeReportGenerate])
	}
	if !IsSourceEventOnly(JobTypeReportGenerate) {
		t.Fatalf("report_generate must be source_event_only")
	}
	if IsSourceEventOnly(JobTypeTargetImport) {
		t.Fatalf("target_import must remain trigger_creates_job")
	}
	if len(JobTriggerEventSemantics) != 7 {
		t.Fatalf("JobTriggerEventSemantics count = %d, want 7", len(JobTriggerEventSemantics))
	}
	for _, internalOnly := range []JobType{JobTypeEmailDispatch} {
		if containsJobType(APIFacingJobTypes, internalOnly) {
			t.Fatalf("%s must stay out of APIFacingJobTypes", internalOnly)
		}
	}
}

func containsJobType(types []JobType, want JobType) bool {
	for _, typ := range types {
		if typ == want {
			return true
		}
	}
	return false
}

File: frontend/src/lib/jobs/jobs.test.ts
import { describe, expect, it } from 'vitest';

import {
  API_FACING_JOB_TYPES,
  ASYNQ_TASK_EMAIL_DISPATCH,
  ASYNQ_TASK_PRIVACY_DELETE,
  ASYNQ_TASK_TARGET_IMPORT,
  isSourceEventOnly,
  JOB_TRIGGER_EVENT_SEMANTIC_SOURCE_EVENT_ONLY,
  JOB_TRIGGER_EVENT_SEMANTIC_TRIGGER_CREATES_JOB,
  JOB_TRIGGER_EVENT_SEMANTICS,
  JOB_TYPE_EMAIL_DISPATCH,
  JOB_TYPE_REPORT_GENERATE,
  JOB_TYPE_SOURCE_REFRESH,
  JOB_TYPE_TARGET_IMPORT,
  EMAIL_DISPATCH_REDACTED_FIELDS,
} from './jobs';

describe('generated job contract', () => {
  it('exports canonical and asynq job names', () => {
    expect(JOB_TYPE_TARGET_IMPORT).toBe('target_import');
    expect(JOB_TYPE_EMAIL_DISPATCH).toBe('email_dispatch');
    expect(ASYNQ_TASK_TARGET_IMPORT).toBe('target.import');
    expect(ASYNQ_TASK_PRIVACY_DELETE).toBe('privacy.delete');
    expect(ASYNQ_TASK_EMAIL_DISPATCH).toBe('email.dispatch');
  });

  it('exports trigger event semantics and source-event predicate', () => {
    expect(JOB_TRIGGER_EVENT_SEMANTIC_SOURCE_EVENT_ONLY).toBe('source_event_only');
    expect(JOB_TRIGGER_EVENT_SEMANTIC_TRIGGER_CREATES_JOB).toBe('trigger_creates_job');
    expect(JOB_TRIGGER_EVENT_SEMANTICS[JOB_TYPE_REPORT_GENERATE]).toBe(JOB_TRIGGER_EVENT_SEMANTIC_SOURCE_EVENT_ONLY);
    expect(isSourceEventOnly(JOB_TYPE_REPORT_GENERATE)).toBe(true);
    expect(isSourceEventOnly(JOB_TYPE_TARGET_IMPORT)).toBe(false);
  });

  it('keeps internal-only jobs out of the API-facing subset', () => {
    expect(API_FACING_JOB_TYPES).toContain('target_import');
    expect(API_FACING_JOB_TYPES).not.toContain(JOB_TYPE_SOURCE_REFRESH);
    expect(API_FACING_JOB_TYPES).not.toContain(JOB_TYPE_EMAIL_DISPATCH);
  });

  it('exports email_dispatch redaction policy', () => {
    expect(EMAIL_DISPATCH_REDACTED_FIELDS).toContain('rawEmailCode');
    expect(EMAIL_DISPATCH_REDACTED_FIELDS).toContain('emailBody');
  });
});
